### PR TITLE
add correctness tracking for experimental parser

### LINKED ISSUE #3503: Add experimental parser correctness checks to tracking
### Describe the feature
`1 / n` parses with the experimental parser should run both python jinja parsing and the experimental parser, compare results and send back whether they match or not. Frequency should be determined with an rng.

### Describe alternatives you've considered
Not having this information

### Additional context
We want to make sure the experimental parser is 100% in line with the existing python jinja rendering before we consider turning it on by default. 

### Who will this benefit?
dbt maintainers

### Relevant Tickets
#3438

### PR DESCRIPTION
**Pull request by nathaniel-may**
resolves #3503

### Description
Adds tracking for experimental parser correctness that includes what kind of mismatch occurred.

### Checklist
 - [x] I have signed the [CLA](https://docs.getdbt.com/docs/contributor-license-agreements)
 - [x] I have run this code in development and it appears to resolve the stated issue
 - [x] This PR includes tests, or tests are not required/relevant for this PR
 - [x] I have updated the `CHANGELOG.md` and added information about my change to the "dbt next" section.

### CODE REVIEW

File: core/dbt/parser/models.py [hunk 1]
```diff
@@ -26,46 +27,109 @@ def render_update(
...
-                # when we support hooks
-                self.update_parsed_node(node, config)
+                    # dedup values
+                    value = list(set(value))
 
-                # udpate the unrendered config with values from the file
-                # values from yaml files are in there already
-                node.unrendered_config.update(dict(res['configs']))
+                    # set sample results
+                    # TODO set this somewhere so it can be sent
```
Comment (jtcohen6) [inline core/dbt/parser/models.py:106]: Right on, @gshank or @kwigley could help with setting up snowplow tracking for this.

We could swing this with just a standard structured event (no schema addition necessary), but I think we'll want a new validation schema that checks for the possible outcome values you've defined above.

Comment (nathaniel-may) [reply core/dbt/parser/models.py:106]: Yeah I have no idea what you might expect to see in the database, so these are just wet noodle values I flung at the wall. What values make sense to you?

**Comment by nathaniel-may** [comment]
Full suite passed with the flag on by default. Reverted flag setting.

File: core/dbt/parser/models.py [hunk 2]
```diff
@@ -26,46 +27,106 @@ def render_update(
...
+        # normal dbt run
+        if not flags.USE_EXPERIMENTAL_PARSER:
+            # normal rendering
+            super().render_update(node, config)
+            # if we're sampling, compare for correctness
+            if sample:
+                value = []
+                # experimental parser couldn't parse
+                if isinstance(experimentally_parsed, Exception):
+                    value += ["01_experimental_parser_couldn't_parse"]
```
Comment (jtcohen6) [inline core/dbt/parser/models.py:61]: Nit: that apostrophe is going to be annoying
```suggestion
                    value += ["01_experimental_parser_cannot_parse"]
```

File: core/dbt/parser/models.py [hunk 3]
```diff
@@ -26,46 +27,106 @@ def render_update(
...
+                        if c not in node.sources:
+                            value += ["04_false_positive_source_value"]
+                            break
 
-                # udpate the unrendered config with values from the file
-                # values from yaml files are in there already
-                node.unrendered_config.update(dict(res['configs']))
+                    # look for missed sources
+                    for c in node.sources:
+                        if c not in experimentally_parsed['configs']:
```
Comment (jtcohen6) [inline core/dbt/parser/models.py:83]:
```suggestion
                        if c not in experimentally_parsed['sources']:
```

File: core/dbt/parser/models.py [hunk 4]
```diff
@@ -26,46 +27,106 @@ def render_update(
...
-                for c in res['configs']:
+                for c in experimentally_parsed['configs']:
                     config_calls.append({c[0]: c[1]})
 
-                config._config_calls = config_calls
+                # format sources TODO change extractor to match this type
+                source_calls = []
+                for s in experimentally_parsed['sources']:
+                    source_calls.append([s[0], s[1]])
+                experimentally_parsed['sources'] = source_calls
```
Comment (jtcohen6) [inline core/dbt/parser/models.py:47]: Need this for `experimentally_parsed['configs'] = config_calls` above as well

File: core/dbt/parser/models.py [hunk 5]
```diff
@@ -26,46 +27,106 @@ def render_update(
...
-                res = py_extract_from_source(node.raw_sql)
+                experimentally_parsed = py_extract_from_source(node.raw_sql)
 
-                # since it doesn't need python jinja, fit the refs, sources, and configs
-                # into the node. Down the line the rest of the node will be updated with
-                # this information. (e.g. depends_on etc.)
+                # second config format
                 config_calls = []
-                for c in res['configs']:
+                for c in experimentally_parsed['configs']:
```
Comment (jtcohen6) [inline core/dbt/parser/models.py:40]: Hmm I'm seeing some misalignments in the sense that the experimental parser returns
```
[{'unique_key': 'event_id'}, {'materialized': 'incremental'}]
```
Whereas traditional parsing returns
```
[{'unique_key': 'event_id', 'materialized': 'incremental'}]
```
With the resulting value of: `['02_false_positive_config_value', '03_missed_config_value']`

File: core/dbt/parser/models.py [hunk 6]
```diff
@@ -26,46 +27,106 @@ def render_update(
...
+                            value += ["07_missed_ref_value"]
+                            break
 
-            # exception was thrown by dbt jinja extractor meaning it can't
-            # handle this source. fall back to python jinja rendering.
-            except ExtractionError:
-                super().render_update(node, config)
+                    # dedup values
+                    value = list(set(value))
+
```
Comment (jtcohen6) [inline core/dbt/parser/models.py:101]: If `value` is empty here at the end, what do you think of sticking in some kind of `"00_success"` or `"00_exact_match"`?

File: core/dbt/parser/models.py (hunk 1 above)
Comment (jtcohen6) [reply core/dbt/parser/models.py:103]: The ones you picked make a lot of sense to me! Just a few tweaks

**Review by jtcohen6 — COMMENTED** [review]
Neat-o! I took this for a spin with our internal analytics project, and dug into some of the discrepancies.

Before merging this in, we need to:
- Instrument a new tracking event
- Run this through all the models in our internal project, and confirm everything comes up exactly matching

File: core/dbt/parser/models.py (hunk 3 above)
Comment (nathaniel-may) [reply core/dbt/parser/models.py:83]: good catch!

File: core/dbt/parser/models.py (hunk 2 above)
Comment (nathaniel-may) [reply core/dbt/parser/models.py:61]: very good point

File: core/dbt/parser/models.py (hunk 4 above)
Comment (nathaniel-may) [reply core/dbt/parser/models.py:47]: configs are weird and need to be in two forms. one for each of the places where we need to set them when the flag is active: `config._config_calls` and `node.unrendered_config.update`. So I purposefully don't overwrite the original format here.

_However_ this correctly points out that my comparison is not using the correct one of the two options so I fixed that.

File: core/dbt/parser/models.py (hunk 5 above)
Comment (nathaniel-may) [reply core/dbt/parser/models.py:40]: yup. I think the above comment should address this now. Thanks for the detailed example!

File: core/dbt/parser/models.py (hunk 6 above)
Comment (nathaniel-may) [reply core/dbt/parser/models.py:101]: yes we should totally do that. added.

File: core/dbt/parser/models.py (hunk 5 above)
Comment (nathaniel-may) [reply core/dbt/parser/models.py:40]: I was wrong, this needed to be handled completely separately. Thanks for pointing it out!

**Comment by nathaniel-may** [comment]
Tested locally with our internal analytics project and it correctly compares for all parseable model files.

**Review by jtcohen6 — COMMENTED** [review]
Thanks for confirming that this matches expectations for our internal project.

Let's wire up a schema + tracking event!

File: core/dbt/parser/models.py [hunk 7]
```diff
@@ -26,46 +27,111 @@ def render_update(
...
+                    # if there are no errors, return a success value
+                    if not result:
+                        result = set(["00_exact_match"])
+
+                    # set sample results. results are cumulative across the dbt run,
+                    # so we will only know there exists a false positive config value
+                    # somewhere in the project, not how many files have false positve
+                    # configs. this means a perfectly correct sample will return a
+                    # result of {'00_exact_match'}.
+                    self.manifest._parsing_info.static_analysis_result.update(result)
```
[the remark this answers is not in the document]
Comment (nathaniel-may) [reply core/dbt/parser/models.py]: just changed it up to fire one event per sampled file.

**Comment by nathaniel-may** [comment]
These CI tests were expected to fail. They will hopefully pass when I uncomment the iglu address once the new schema has been added.

**Comment by nathaniel-may** [comment]
rebased on branch `test-tweaks` to apply changes here before `test-tweaks` gets merged into develop. Only reason is to continue working while Circle is having issues.

File: core/dbt/parser/models.py [hunk 8]
```diff
@@ -26,46 +31,122 @@ def render_update(
...
+                    # fire a tracking event. this fires one event for every sample
+                    # so that we have data on a per file basis. Not only can we expect
+                    # no false positives or misses, we can expect the number model
+                    # files parseable by the experimental parser to match our internal
+                    # testing.
+                    tracking.track_experimental_parser_sample({
+                        "project_id": self.root_project.hashed_name(),
+                        "file_id": utils.get_hash(node),
+                        "status": result
+                    })
```
Comment (jtcohen6) [inline core/dbt/parser/models.py:123]: Let's also track if the result is `01_experimental_parser_cannot_parse`!

**Comment by nathaniel-may** [comment]
closing in favor of [#3553](https://github.com/dbt-labs/dbt/pull/3553)

**Pull request closed without merging**